### PR TITLE
Simplex segmented container

### PR DESCRIPTION
**Pull request by biancagi**
This branch integrates the segmented container within the design of simplex factories. 

Now, `OrientedSimplexFactory` and `SimplexFactory` layout the simplices that they own in `SegmentedContainer` data structures, as opposed to creating new simplices in scattered memory locations.

### CODE REVIEW

File: lib/mito/topology/OrientedSimplexFactory.h [hunk 1]
```diff
@@ -180,63 +200,44 @@ namespace mito::topology {
...
     class OrientedSimplexFactory<0> {
 
+        // typedef for a collection of vertices
+        using vertex_collection_t = element_collection_t<vertex_t>;
+
       public:
-        // delete default constructor
-        OrientedSimplexFactory() = delete;
+        // default constructor
+        OrientedSimplexFactory() : _vertices(100 /*segment size */) {};
```
Comment (akoshakji) [inline lib/mito/topology/OrientedSimplexFactory.h:214]: same comment for the hard coded size

File: lib/mito/utilities/Shareable.h [hunk 2]
```diff
@@ -42,20 +39,22 @@ namespace mito::utilities {
...
       private:
-        int _reference_count;
+        mutable int _reference_count;
 
       private:
         // friendship with SharedPointer (the shared pointer needs r/w access to the reference count
         // of the Shareable instance)
         template <class T>
-        requires mito::utilities::ReferenceCountedObject<T>
+        // requires mito::utilities::ReferenceCountedObject<T>
```
Comment (akoshakji) [inline lib/mito/utilities/Shareable.h:57]: remove commented code if not needed

File: lib/mito/utilities/SharedPointer.h [hunk 3]
```diff
@@ -11,7 +11,7 @@ namespace mito::utilities {
 
     // declaration
     template <class Resource>
-    requires ReferenceCountedObject<Resource>
+    // requires ReferenceCountedObject<Resource>
```
Comment (akoshakji) [inline lib/mito/utilities/SharedPointer.h:14]: remove commented code if not needed

File: lib/mito/utilities/SharedPointer.h [hunk 4]
```diff
@@ -32,6 +32,9 @@ namespace mito::utilities {
         // operator->
         auto operator->() const -> handle_t;
 
+        // // operator*
+        // auto operator*() const -> const resource_t &;
```
Comment (akoshakji) [inline lib/mito/utilities/SharedPointer.h:36]: remove

File: lib/mito/utilities/forward.h [hunk 5]
```diff
@@ -14,15 +14,15 @@ namespace mito::utilities {
 
     // class shared pointer based on a reference counted resource
     template <class Resource>
-    requires ReferenceCountedObject<Resource>
+    // requires ReferenceCountedObject<Resource>
```
Comment (akoshakji) [inline lib/mito/utilities/forward.h:17]: remove also here

**Review by akoshakji — COMMENTED** [review]
Some minor comments. Great work! This PR fixes #15 #22 #27 #28

File: lib/mito/geometry/PointCloud.h [hunk 6]
```diff
@@ -7,60 +7,54 @@ namespace mito::geometry {
     template <int D>
     class PointCloud {
       private:
-        using cloud_t = std::unordered_map<vertex_t, point_t<D>>;
+        using cloud_t = mito::geometry::cloud_t<D>;
 
       public:
-        PointCloud() = delete;
+        PointCloud() : _cloud(100 /*segment size */) {}
```
[the remark this answers is not in the document]
Comment (biancagi) [reply lib/mito/geometry/PointCloud.h:13]: The optimal choice of the segment size for a segmented container is expected to depend strongly on the problem size. Whether that information is selected by the user or computed automatically in the library is still to be determined. This holds for `PointCloud` as well as for `SimplexFactory` and `OrientedSimplexFactory`. I have opened an issue on this topic (see issue #33).

File: lib/mito/geometry/PointCloud.h [hunk 7]
```diff
@@ -7,60 +7,54 @@ namespace mito::geometry {
...
 
       public:
-        static auto print() -> void
+        auto print() const -> void
```
[the remark this answers is not in the document]
Comment (biancagi) [reply lib/mito/geometry/PointCloud.h:16]: This is not specific of this pull request and should be addressed in a broader context. I have opened a new issue on this topic (see issue #34).

File: lib/mito/manifolds/metrics.h [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+    template <int D>
+    auto computeSimplicesVolume(
+        const element_vector_t<topology::simplex_t<D>> & elements,
+        const mesh::vertex_point_table_t<D> & vertices, std::vector<real> & volumes) -> void
+    {
+        // number of element vertices
+        constexpr int V = D + 1;
+
+        // a container to store the coordinates of each vertex in a tensor
+        static matrix_t<V> pointsTensor;
```
[the remark this answers is not in the document]
Comment (biancagi) [reply lib/mito/manifolds/metrics.h]: Removed {static} declaration.

File: lib/mito/manifolds/metrics.h [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+                    pointsTensor[v * V + d] = vertices.find(vertex)->second->coordinates()[d];
+                }
+                pointsTensor[v * V + D] = 1.0;
+                // update element vertices counter
+                ++v;
+            }
+
+            // compute the volume of the e-th element
+            volumes[e] =
+                fabs(pyre::algebra::determinant(pointsTensor)) / pyre::algebra::factorial<D>();
```
[the remark this answers is not in the document]
Comment (biancagi) [reply lib/mito/manifolds/metrics.h]: Done!

File: lib/mito/utilities/SharedPointer.h (hunk 3 above)
Comment (biancagi) [reply lib/mito/utilities/SharedPointer.h]: The {requires} statement has now been reinstated.

File: lib/mito/utilities/Shareable.h (hunk 2 above)
Comment (biancagi) [reply lib/mito/utilities/Shareable.h]: The {requires} statement has been reinstated.

File: lib/mito/utilities/SharedPointer.h (hunk 4 above)
Comment (biancagi) [reply lib/mito/utilities/SharedPointer.h:36]: I'd rather leave it as we may want to implement {operator*} for shared pointer in the future.

**Review by akoshakji — APPROVED** [review]

**Comment by biancagi** [comment]
Fixes #15, #22, #27, #28.

**Pull request merged**